### PR TITLE
reformat build suffix so that the latest is always correct

### PR DESCRIPTION
**Pull request by liqunfu**
**Description**:         
# we want to format the build_suffix to avoid (the 12th run on 20210630 vs the first run on 20210701):
        # 2021063012 > 20210701
        # in above 2021063012 is treated as the latest which is incorrect.
        # we want to convert the format to:
        # 20210630012 < 20210701001
        # where the first 8 digits are date. the last 3 digits are run count.
        # as long as there are less than 1000 runs per day, we will not have the problem.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -383,9 +383,40 @@ def get_torch_version():
...
+                    return "{:03}".format(count)
+                return ""
+            except: # noqa
+                return ""
+
+        build_suffix_is_date_format = check_date_format(build_suffix[:8])
+        build_suffix_run_count = reformat_run_count(build_suffix[8:])
+        if build_suffix_is_date_format and build_suffix_run_count:
+            build_suffix = build_suffix[:8] + build_suffix_run_count
+
```
Comment (thiagocrepaldi) [inline setup.py:419]:
```suggestion
    elif len(build_suffix) >= 12:
        raise RuntimeError(f'Incorrect build suffix: "{build_suffix}"')

```

File: setup.py [hunk 2]
```diff
@@ -383,9 +383,40 @@ def get_torch_version():
...
+                return True
+            except: # noqa
+                return False
+
+        def reformat_run_count(count_str):
+            try:
+                count = int(count_str)
+                if count >= 0 and count < 1000:
+                    return "{:03}".format(count)
+                return ""
```
Comment (thiagocrepaldi) [inline setup.py:411]:
```suggestion
                elif count >= 1000:
                    raise RuntimeError(f'Too many builds for the same day: {count}')
                return ""
```
Not entirely needed due to the build_suffix length above

File: setup.py [hunk 3]
```diff
@@ -383,9 +383,40 @@ def get_torch_version():
     if build_suffix is None:
         # The following line is only for local testing
         build_suffix = str(datetime.datetime.now().date().strftime("%Y%m%d"))
+        print("build_suffix1: ", build_suffix)
     else:
+        print("build_suffix2: ", build_suffix)
```
Comment (thiagocrepaldi) [inline setup.py:388]:
```suggestion
```

File: setup.py [hunk 4]
```diff
@@ -383,9 +383,40 @@ def get_torch_version():
     if build_suffix is None:
         # The following line is only for local testing
         build_suffix = str(datetime.datetime.now().date().strftime("%Y%m%d"))
+        print("build_suffix1: ", build_suffix)
```
Comment (thiagocrepaldi) [inline setup.py:386]:
```suggestion
```

File: setup.py [hunk 5]
```diff
@@ -295,7 +295,7 @@ def run(self):
     package_name = 'onnxruntime-training'
 
     # we want put default training packages to pypi. pypi does not accept package with a local version.
-    if not default_training_package_device:
+    if (not default_training_package_device) or nightly_build:
```
Comment (thiagocrepaldi) [inline setup.py:298]:
```suggestion
    if not default_training_package_device or nightly_build:
```

**Review by thiagocrepaldi — APPROVED** [review]

File: setup.py (hunk 2 above)
Comment (liqunfu) [reply setup.py:411]: nice

**Review by suffiank — APPROVED** [review]
Great! This is needed.

**Review by baijumeswani — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]

**Pull request merged**